README.md
# NAME

App-gimpgitbuild - build GIMP from git

# VERSION

version 0.10.1

# SYNOPSIS

    gimpgitbuild build

# DESCRIPTION

gimpgitbuild is a command line utility to automatically build
[GIMP](https://www.gimp.org/) (= the "GNU Image Manipulation Program")
and some of its dependencies from its version control git repositories:
[https://developer.gimp.org/git.html](https://developer.gimp.org/git.html) .

Use it only if your paths and environment does not contain too many
nasty characters because we interpolate strings into the shell a lot.

So far, it is quite opinionated, but hopefully we'll allow for better
customization using [https://en.wikipedia.org/wiki/Environment\_variable](https://en.wikipedia.org/wiki/Environment_variable)
in the future.

## HISTORY

This utility evolved from an [old bash version](https://github.com/shlomif/shlomif-computer-settings/blob/db468a5d6190bce053af1621b30e7dfd673be43f/shlomif-settings/build-scripts/build/gimp-git-all-deps.bash)
and a [rewrite in perl 5](https://github.com/shlomif/shlomif-computer-settings/blob/master/shlomif-settings/build-scripts/build/gimp-git-all-deps.pl) .

# SEE ALSO

- [https://www.gimp.org/](https://www.gimp.org/)

# SUPPORT

## Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

- MetaCPAN

    A modern, open-source CPAN search engine, useful to view POD in HTML format.

    [https://metacpan.org/release/App-gimpgitbuild](https://metacpan.org/release/App-gimpgitbuild)

- RT: CPAN's Bug Tracker

    The RT ( Request Tracker ) website is the default bug/issue tracking system for CPAN.

    [https://rt.cpan.org/Public/Dist/Display.html?Name=App-gimpgitbuild](https://rt.cpan.org/Public/Dist/Display.html?Name=App-gimpgitbuild)

- CPANTS

    The CPANTS is a website that analyzes the Kwalitee ( code metrics ) of a distribution.

    [http://cpants.cpanauthors.org/dist/App-gimpgitbuild](http://cpants.cpanauthors.org/dist/App-gimpgitbuild)

- CPAN Testers

    The CPAN Testers is a network of smoke testers who run automated tests on uploaded CPAN distributions.

    [http://www.cpantesters.org/distro/A/App-gimpgitbuild](http://www.cpantesters.org/distro/A/App-gimpgitbuild)

- CPAN Testers Matrix

    The CPAN Testers Matrix is a website that provides a visual overview of the test results for a distribution on various Perls/platforms.

    [http://matrix.cpantesters.org/?dist=App-gimpgitbuild](http://matrix.cpantesters.org/?dist=App-gimpgitbuild)

- CPAN Testers Dependencies

    The CPAN Testers Dependencies is a website that shows a chart of the test results of all dependencies for a distribution.

    [http://deps.cpantesters.org/?module=App::gimpgitbuild](http://deps.cpantesters.org/?module=App::gimpgitbuild)

## Bugs / Feature Requests

Please report any bugs or feature requests by email to `bug-app-gimpgitbuild at rt.cpan.org`, or through
the web interface at [https://rt.cpan.org/Public/Bug/Report.html?Queue=App-gimpgitbuild](https://rt.cpan.org/Public/Bug/Report.html?Queue=App-gimpgitbuild). You will be automatically notified of any
progress on the request by the system.

## Source Code

The code is open to the world, and available for you to hack on. Please feel free to browse it and play
with it, or whatever. If you want to contribute patches, please send me a diff or prod me to pull
from your repository :)

[https://github.com/shlomif/App-gimpgitbuild](https://github.com/shlomif/App-gimpgitbuild)

    git clone git://github.com/shlomif/App-gimpgitbuild.git

# AUTHOR

Shlomi Fish <shlomif@cpan.org>

# BUGS

Please report any bugs or feature requests on the bugtracker website
[https://github.com/shlomif/App-gimpgitbuild/issues](https://github.com/shlomif/App-gimpgitbuild/issues)

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2020 by Shlomi Fish.

This is free software, licensed under:

    The Apache License, Version 2.0, January 2004


App-gimpgitbuild/Changes
Revision history for App-gimpgitbuild

0.30.1   2021-10-10
    - install gimp before testing it.
        - Workaround.

0.30.0   2021-09-19
    - Build and install gexiv2 as well.
        - shlomif has volunteered to comaintain gexiv2:
        - https://mail.gnome.org/archives/gimp-developer-list/2021-September/msg00004.html

0.28.1   2021-08-30
    - Restrict clean_install to the "build" mode.

0.28.0   2021-08-30
    - Disable lua in GEGL and gimp:
        - https://gitlab.gnome.org/GNOME/gimp/-/issues/7196
    - Delete the installation directories to avoid stale files
    - Apply LD_LIBRARY_PATH

0.26.1   2020-07-28
    - Avoid recommending against a meson build when the mode is "non-build"
    (e.g "clean" mode).
        - It is irrelevant to it.

0.26.0   2020-06-22
    - Refactoring and cleanups.

0.24.0   2020-06-22
    - Add the "gimpgitbuild cleanbuild" sub-command.
        - It is an alias to "gimpgitbuild build --mode=clean"
        which will keep working for backwards compatibility.
    - Extract an API::Worker class to facilitate the "cleanbuild"
    change.

0.22.0   2020-06-20
    - Add the --process-exe flag to gimpgitbuild build
        - It defaults to "perl" which is the new behavior.
    - Cleanups, code refactoring, and sanity checks.

0.20.0   2020-06-20
    - Add a note for failed meson-using builds of GIMP.
    - Refactoring and cleanups.

0.18.1   2020-06-09
    - Make sure an optional gvim invocation does not emit gtk warnings.
        - They break gimp's "make check".

0.18.0   2020-06-03
    - Add "build --mode=clean"
    - Build libmypaint tag version 1.6.1.

0.16.2   2020-05-13
    - Some code cleanups and refactorings.
        - https://en.wikipedia.org/wiki/Code_refactoring

0.16.1   2020-04-02
    - Require xvfb-run
    - Add --enable-debug to gimp build
    - https://gitlab.gnome.org/GNOME/gimp/-/issues/2884

0.16.0   2020-03-29
    - Add the GIMPGITBUILD__BUILD_GIMP_USING_MESON env var.

0.14.0   2020-03-25
    - Avoid calling the meson build directories "build" and allow
    for customisation.
    - Add the PAR_JOBS customisation.

0.12.1   2020-03-19
    - Avoid "undefined" warning for an environment variable.

0.12.0   2020-03-08
    - Add the eval "$(gimpgitbuild runenv)" sub-command for LD_LIBRARY_PATH
    runtime vars.
        - By inspiration from https://mail.gnome.org/archives/gimp-developer-list/2020-February/msg00026.html
        - Thanks, Owen!

0.10.2   2020-03-05
    - Resubmit due to a PAUSE indexing failure; trying to get the new version
    indexed.

0.10.1   2020-03-02
    - Fix missing $VERSION in lib/**.pm files.

0.10.0   2020-03-02
    - Use the API's class' extracted methods.
    - Remove some unused copy+pasted code.

0.8.0   2020-02-29
    - Add the GIMPGITBUILD__SKIP_BUILDS_RE environment variable

0.6.1   2020-02-28
    - Fix meson installation libdir in Ubuntu and derivatives.
        - Thanks to Owen for the report
        - Thanks to https://github.com/libfuse/libfuse/issues/212 for the fix.

0.6.0   2020-02-27
    - Use Git::Sync::App instead of git-s.
        - git-s may not be installed by default.
        - See https://mail.gnome.org/archives/gimp-developer-list/2020-February/msg00016.html
        - Thanks to Owen

0.4.0   2020-02-25
    - Support GIMPGITBUILD__BASE_INSTALL_DIR
    and GIMPGITBUILD__BASE_CLONES_DIR env vars
    - Add docs.

0.2.1   2020-02-25
    - Fix "env"'s description.

0.2.0   2020-02-25
    - Add the "env" subcommand.

0.0.2   2020-02-18
    - Rebuild with updated deps to avoid old links in generated POD.

0.0.1   2020-02-18
    - First version, released on an unsuspecting world


App-gimpgitbuild/lib/App/gimpgitbuild.pm
# ABSTRACT: gimp build
package App::gimpgitbuild;

use strict;
use warnings;

use App::Cmd::Setup -app;

1;

__END__

=head1 NAME

App-gimpgitbuild - build GIMP from git

=head1 SYNOPSIS

    gimpgitbuild build

=head1 DESCRIPTION

gimpgitbuild is a command line utility to automatically build
L<GIMP|https://www.gimp.org/> (= the "GNU Image Manipulation Program")
and some of its dependencies from its version control git repositories:
L<https://developer.gimp.org/git.html> .

Use it only if your paths and environment does not contain too many
nasty characters because we interpolate strings into the shell a lot.

So far, it is quite opinionated, but hopefully we'll allow for better
customization using L<https://en.wikipedia.org/wiki/Environment_variable>
in the future.

=head2 HISTORY

This utility evolved from an L<old bash version|https://github.com/shlomif/shlomif-computer-settings/blob/db468a5d6190bce053af1621b30e7dfd673be43f/shlomif-settings/build-scripts/build/gimp-git-all-deps.bash>
and a L<rewrite in perl 5|https://github.com/shlomif/shlomif-computer-settings/blob/master/shlomif-settings/build-scripts/build/gimp-git-all-deps.pl> .

=head1 SEE ALSO

=over 4

=item * L<https://www.gimp.org/>

=back

=cut


App-gimpgitbuild/lib/App/gimpgitbuild/Command/env.pm
package App::gimpgitbuild::Command::env;

use strict;
use warnings;
use 5.014;

use App::gimpgitbuild -command;

use Path::Tiny qw/ path tempdir tempfile cwd /;

use App::gimpgitbuild::API::GitBuild ();

sub description
{
    return "set the environment for using or building GIMP-from-git";
}

sub abstract
{
    return shift->description();
}

sub opt_spec
{
    return ();

=begin foo
    return (
        [ "output|o=s", "Output path" ],
        [ "title=s",    "Chart Title" ],
        [ 'exec|e=s@',  "Execute command on the output" ]
    );
=end foo

=cut

}

sub execute
{
    my ( $self, $opt, $args ) = @_;

    my $output_fn = $opt->{output};
    my $exe       = $opt->{exec} // [];

    my $obj = App::gimpgitbuild::API::GitBuild->new;

    my $env = $obj->new_env;
    print <<"EOF";
export PATH="$env->{PATH}" ;
export PKG_CONFIG_PATH="$env->{PKG_CONFIG_PATH}" ;
export XDG_DATA_DIRS="$env->{XDG_DATA_DIRS}" ;
EOF

    return;
}

1;

__END__

=head1 NAME

gimpgitbuild env - set the preferred environment in the shell.

=head1 SYNOPSIS

    # In your sh-compatible shell:
    eval "$(gimpgitbuild env)"

=cut


App-gimpgitbuild/lib/App/gimpgitbuild/API/GitBuild.pm
package App::gimpgitbuild::API::GitBuild;

use strict;
use warnings;
use 5.014;

use Moo;

has home_dir         => ( is => 'lazy' );
has install_base_dir => ( is => 'lazy' );

sub _build_home_dir
{
    return $ENV{HOME};
}

sub _build_install_base_dir
{
    my $self = shift;
    return $ENV{GIMPGITBUILD__BASE_INSTALL_DIR}
        // ( $self->home_dir . "/apps/graphics" );
}

sub mypaint_p
{
    my $self = shift;
    return $self->install_base_dir . "/libmypaint";
}

sub babl_p
{
    my $self = shift;
    return $self->install_base_dir . "/babl";
}

sub gimp_p
{
    my $self = shift;
    return $self->install_base_dir . "/gimp-devel";
}

sub gegl_p
{
    my $self = shift;
    return $self->install_base_dir . "/gegl";
}

sub gexiv2_p
{
    my $self = shift;
    return $self->gegl_p();
}

sub base_git_clones_dir
{
    my $self = shift;

    return $ENV{GIMPGITBUILD__BASE_CLONES_DIR}
        // ( $self->home_dir . "/Download/unpack/graphics/gimp" );
}

sub new_env
{
    my $self            = shift;
    my $gegl_p          = $self->gegl_p;
    my $gimp_p          = $self->gimp_p;
    my $babl_p          = $self->babl_p;
    my $mypaint_p       = $self->mypaint_p;
    my $PKG_CONFIG_PATH = join(
        ":",
        (
            map {
                my $p = $_;
                map { "$p/$_/pkgconfig" } qw# share lib64 lib  #
            } ( $babl_p, $gegl_p, $mypaint_p )
        ),
        ( $ENV{PKG_CONFIG_PATH} // '' )
    );
    my $LD_LIBRARY_PATH = join(
        ":",
        (
            map {
                my $p = $_;
                map { "$p/$_" } qw# lib64 lib  #
            } ( $gimp_p, $babl_p, $gegl_p, $mypaint_p )
        ),
        ( $ENV{LD_LIBRARY_PATH} // '' )
    );
    my $xdg_prefix =
"$gegl_p/share:$mypaint_p/share:$mypaint_p/share/pkgconfig:$babl_p/share";
    my $XDG_DATA_DIRS = (
        exists( $ENV{XDG_DATA_DIRS} )
        ? "$xdg_prefix:$ENV{XDG_DATA_DIRS}"
        : $xdg_prefix
    );
    return +{
        LD_LIBRARY_PATH => $LD_LIBRARY_PATH,
        PATH            => "$gegl_p/bin:$ENV{PATH}",
        PKG_CONFIG_PATH => $PKG_CONFIG_PATH,
        XDG_DATA_DIRS   => $XDG_DATA_DIRS,
    };
}

1;

__END__

=head1 NAME

App::gimpgitbuild::API::GitBuild - common API

=head1 METHODS

=head2 babl_p

The BABL install prefix.

=head2 gegl_p

The GEGL install prefix.

=head2 gexiv2_p

The gexiv2 install prefix.

=head2 gimp_p

The GIMP install prefix.

=head2 mypaint_p

The libmypaint install prefix.

=head2 new_env

Returns a hash reference of new environment variables to override.

=head2 install_base_dir

Can be overrided by setting the C<GIMPGITBUILD__BASE_INSTALL_DIR> environment
variable.

=head2 base_git_clones_dir

The base filesystem directory path for the git repository clones.
Can be overrided by setting the C<GIMPGITBUILD__BASE_CLONES_DIR> environment
variable.

=cut


App-gimpgitbuild/lib/App/gimpgitbuild/Command/build.pm
package App::gimpgitbuild::Command::build;

use strict;
use warnings;
use autodie;
use 5.014;

use App::gimpgitbuild -command;

use File::Which qw/ which /;

use App::gimpgitbuild::API::GitBuild ();
use App::gimpgitbuild::API::Worker   ();
use Git::Sync::App                   ();

sub description
{
    return "build gimp from git";
}

sub abstract
{
    return shift->description();
}

sub opt_spec
{
    return ( [ "mode=s", "Mode (e.g: \"clean\")" ],
        [ "process-exe=s", qq#Process executor (= "sh" or "perl")#, ] );

=begin foo
    return (
        [ "output|o=s", "Output path" ],
        [ "title=s",    "Chart Title" ],
        [ 'exec|e=s@',  "Execute command on the output" ]
    );
=end foo

=cut

}

sub _ascertain_xvfb_run_presence
{
    my $path = which('xvfb-run');
    if ( not defined($path) )
    {
        die
"Cannot find xvfb-run ! It is required for tests to succeed: see https://gitlab.gnome.org/GNOME/gimp/-/issues/2884";
    }
    return;
}

sub _ascertain_lack_of_gtk_warnings
{
    my $path = which('gvim');
    if ( defined($path) )
    {
        my $stderr = `"$path" -u NONE -U NONE -f /dev/null +q 2>&1`;
        if ( $stderr =~ /\S/ )
        {
            die
"There may be gtk warnings (e.g: in KDE Plasma 5 on Fedora 32 ). Please fix them.";
        }
    }
    return;
}

sub _ascertain_gjs_presence
{
    my $path = which('gjs');
    if ( not defined($path) )
    {
        die
"gjs must be present for GIMP's tests to succeed - please install it (see: https://gitlab.gnome.org/GNOME/gimp/-/issues/7341 )";
    }
    return;
}

sub execute
{
    my ( $self, $opt, $args ) = @_;

    my $mode = ( $opt->{mode} || 'build' );
    if ( not( ( $mode eq 'clean' ) or ( $mode eq 'build' ) ) )
    {
        die "Unsupported mode '$mode'!";
    }

    my $_process_executor = ( $opt->{process_exe} || 'perl' );
    if (
        not(   ( $_process_executor eq 'sh' )
            or ( $_process_executor eq 'perl' ) )
        )
    {
        die "Unsupported process-exe '$_process_executor'!";
    }

    my $worker = App::gimpgitbuild::API::Worker->new(
        { _mode => $mode, _process_executor => $_process_executor, } );

    my $env = App::gimpgitbuild::API::GitBuild->new()->new_env();
    $ENV{LD_LIBRARY_PATH} = $env->{LD_LIBRARY_PATH};
    $ENV{PATH}            = $env->{PATH};
    $ENV{PKG_CONFIG_PATH} = $env->{PKG_CONFIG_PATH};
    $ENV{XDG_DATA_DIRS}   = $env->{XDG_DATA_DIRS};
    _ascertain_xvfb_run_presence();
    _ascertain_lack_of_gtk_warnings();
    _ascertain_gjs_presence();

    $worker->_run_the_mode_on_all_repositories();

    use Term::ANSIColor qw/ colored /;
    print colored( [ $ENV{HARNESS_SUMMARY_COLOR_SUCCESS} || 'bold green' ],
        "\n== Success ==\n\n" );
    return;
}

1;

__END__

=head1 NAME

gimpgitbuild build - command line utility to automatically build GIMP and its dependencies from git.

=cut


App-gimpgitbuild/lib/App/gimpgitbuild/Command/runenv.pm
package App::gimpgitbuild::Command::runenv;

use strict;
use warnings;
use 5.014;

use App::gimpgitbuild -command;

use Path::Tiny qw/ path tempdir tempfile cwd /;

use App::gimpgitbuild::API::GitBuild ();

sub description
{
    return "set the environment for running GIMP-from-git";
}

sub abstract
{
    return shift->description();
}

sub opt_spec
{
    return ();

=begin foo
    return (
        [ "output|o=s", "Output path" ],
        [ "title=s",    "Chart Title" ],
        [ 'exec|e=s@',  "Execute command on the output" ]
    );
=end foo

=cut

}

sub execute
{
    my ( $self, $opt, $args ) = @_;

    my $output_fn = $opt->{output};
    my $exe       = $opt->{exec} // [];

    my $obj = App::gimpgitbuild::API::GitBuild->new;

    my $env = $obj->new_env;
    print <<"EOF";
export LD_LIBRARY_PATH="$env->{LD_LIBRARY_PATH}" ;
EOF

    return;
}

1;

__END__

=head1 NAME

gimpgitbuild runenv - set the environment in the shell for running the gimp-from-git
installation.

=head1 SYNOPSIS

    # In your sh-compatible shell:
    eval "$(gimpgitbuild runenv)"

=cut


App-gimpgitbuild/lib/App/gimpgitbuild/Command/cleanbuild.pm
package App::gimpgitbuild::Command::cleanbuild;

use strict;
use warnings;
use autodie;
use 5.014;

use App::gimpgitbuild -command;

use App::gimpgitbuild::API::Worker ();

sub description
{
    return "clean the GIMP build checkouts";
}

sub abstract
{
    return shift->description();
}

sub opt_spec
{
    return ();

=begin foo
    return (
        [ "output|o=s", "Output path" ],
        [ "title=s",    "Chart Title" ],
        [ 'exec|e=s@',  "Execute command on the output" ]
    );
=end foo

=cut

}

sub execute
{
    my ( $self, $opt, $args ) = @_;

    my $worker = App::gimpgitbuild::API::Worker->new(
        { _mode => 'clean', _process_executor => 'perl', } );

    $worker->_run_the_mode_on_all_repositories();

    use Term::ANSIColor qw/ colored /;
    print colored( [ $ENV{HARNESS_SUMMARY_COLOR_SUCCESS} || 'bold green' ],
        "\n== Success ==\n\n" );
    return;
}

1;

__END__

=head1 NAME

gimpgitbuild cleanbuild - clean the build checkouts / working copies.

=cut


App-gimpgitbuild/lib/App/gimpgitbuild/API/Worker.pm
package App::gimpgitbuild::API::Worker;

use strict;
use warnings;
use 5.014;

use Moo;

use Path::Tiny qw/ path cwd /;
use Git::Sync::App                   ();
use App::gimpgitbuild::API::GitBuild ();

has '_api_obj' => (
    is      => 'rw',
    default => sub { return App::gimpgitbuild::API::GitBuild->new(); }
);
has '_mode'             => ( is => 'ro', required => 1, );
has '_override_mode'    => ( is => 'rw', default  => "", );
has '_process_executor' => ( is => 'ro', required => 1, );

sub _do_system
{
    my ($args) = @_;

    my $cmd = $args->{cmd};
    print "Running [@$cmd]\n";
    if ( system(@$cmd) )
    {
        die "Running [@$cmd] failed!";
    }
}

my $PAR_JOBS = ( $ENV{GIMPGITBUILD__PAR_JOBS_FLAGS} // '-j4' );
my $skip_builds_re;

BEGIN
{
    my $KEY = "GIMPGITBUILD__SKIP_BUILDS_RE";
    if ( exists $ENV{$KEY} )
    {
        my $re_str = $ENV{$KEY};
        $skip_builds_re = qr/$re_str/;
    }
}
my $BUILD_DIR = ( $ENV{GIMPGITBUILD__MESON_BUILD_DIR}
        // "to-del--gimpgitbuild--build-dir" );

# See:
# https://github.com/libfuse/libfuse/issues/212
# Ubuntu/etc. places it under $prefix/lib/$arch by default.
my $UBUNTU_MESON_LIBDIR_OVERRIDE = "-D libdir=lib";

sub _check
{
    return ( length( $ENV{SKIP_CHECK} ) ? "true" : "make check" );
}

sub _git_sync
{
    my ( $self, $args ) = @_;
    return
qq#$^X -MGit::Sync::App -e "Git::Sync::App->new->run" -- sync origin "$args->{branch}"#;
}

sub _git_build
{
    my $self                 = shift;
    my $args                 = shift;
    my $orig_cwd             = cwd()->absolute();
    my $id                   = $args->{id};
    my $extra_configure_args = ( $args->{extra_configure_args} // [] );
    my $extra_meson_args     = ( $args->{extra_meson_args}     // [] );
    my $SHELL_PREFIX         = "set -e -x";
    my $install_before_test  = ( $args->{install_before_test} // '' );

    if ( defined($skip_builds_re) and $id =~ $skip_builds_re )
    {
        return;
    }
    $args->{branch} //= 'master';
    $args->{tag}    //= 'false';

    my $git_co = (
        $args->{git_co} // (
                  $self->_api_obj()->base_git_clones_dir() . "/"
                . $args->{git_checkout_subdir}
        )
    );
    if ( !-e $git_co )
    {
        path($git_co)->parent->mkpath;
        _do_system( { cmd => [qq#git clone "$args->{url}" "$git_co"#] } );
    }

    my $_autodie_chdir = sub {
        my $dirname = shift;
        if ( not chdir($dirname) )
        {
            die qq#Failed changing directory to "$dirname"!#;
        }
        return;
    };
    my $shell_cmd = sub {
        return shift;
    };
    my $chdir_cmd = sub {
        return $shell_cmd->( qq#cd "# . shift(@_) . qq#"# );
    };
    my $clean_install = ( $self->_override_mode() eq "clean_install" );
    my $PERL_EXECUTE =
        ( $clean_install or $self->_process_executor() eq 'perl' );
    if ($PERL_EXECUTE)
    {
        $shell_cmd = sub {
            my $cmd = shift;
            return sub {
                return _do_system(
                    {
                        cmd => ["$SHELL_PREFIX ; $cmd"],
                    }
                );
            };
        };
        $chdir_cmd = sub {
            my $dirname = shift;
            return sub {
                return $_autodie_chdir->($dirname);
            };
        };
    }

    my $prefix = $args->{prefix};

    if ($clean_install)
    {
        $shell_cmd->(qq#rm -fr "$prefix"#)->();
        return;
    }

    my $test_install_order = sub {
        my ( $test_cmd, $install_cmd ) = @_;
        return $install_before_test
            ? [ @$install_cmd, @$test_cmd ]
            : [ @$test_cmd, @$install_cmd ];
    };

    my $gen_meson_build_cmds = sub {
        return [
            $shell_cmd->(qq#mkdir -p "$BUILD_DIR"#),
            $chdir_cmd->($BUILD_DIR),
            $shell_cmd->(
qq#meson --prefix="$prefix" $UBUNTU_MESON_LIBDIR_OVERRIDE @{$extra_meson_args} ..#
            ),
            $shell_cmd->(qq#ninja $PAR_JOBS#),
            @{
                $test_install_order->(
                    [ $shell_cmd->(qq#ninja $PAR_JOBS test#), ],
                    [ $shell_cmd->(qq#ninja $PAR_JOBS install#), ]
                )
            },
        ];
    };
    my $gen_autoconf_build_cmds = sub {
        return [
            $shell_cmd->(qq#NOCONFIGURE=1 ./autogen.sh#),
            $shell_cmd->(qq#mkdir -p "$BUILD_DIR"#),
            $chdir_cmd->($BUILD_DIR),
            $shell_cmd->(
                qq#../configure @{$extra_configure_args} --prefix="$prefix"#),
            $shell_cmd->(qq#make $PAR_JOBS#),
            @{
                $test_install_order->(
                    [ $shell_cmd->(qq#@{[_check()]}#), ],
                    [ $shell_cmd->(qq#make install#), ]
                )
            },
        ];
    };
    my $gen_clean_mode_cmds =
        sub { return [ $shell_cmd->(qq#git clean -dxf .#), ]; };
    my $commands_gen = (
        ( $self->_mode() eq 'clean' ) ? $gen_clean_mode_cmds
        : (
              $args->{use_meson} ? $gen_meson_build_cmds
            : $gen_autoconf_build_cmds
        )
    );
    my $sync_cmd = $self->_git_sync( { branch => $args->{branch}, } );
    my @commands = (
        $chdir_cmd->($git_co),
        $shell_cmd->(qq#git checkout "$args->{branch}"#),
        $shell_cmd->(qq#( $args->{tag} || $sync_cmd )#),
        @{ $commands_gen->() },
    );

    my $run = sub {
        if ($PERL_EXECUTE)
        {
            foreach my $cb (@commands)
            {
                $cb->();
            }
            return;
        }
        my $aggregate_shell_command =
            "$SHELL_PREFIX ; " . join( " ; ", @commands );
        return _do_system(
            {
                cmd => [ $aggregate_shell_command, ]
            }
        );
    };

    my $on_failure = $args->{on_failure};

    if ( !$on_failure )
    {
        $run->();
    }
    else
    {
        eval { $run->(); };
        my $Err = $@;

        if ($Err)
        {
            $on_failure->( { exception => $Err, }, );
        }
    }
    $_autodie_chdir->($orig_cwd);
    return;
}

sub _get_gnome_git_url
{
    my ( $self, $proj ) = @_;
    my $GNOME_GIT = 'https://gitlab.gnome.org/GNOME';

    return "${GNOME_GIT}/${proj}.git/";
}

sub _run_all
{
    my ($worker) = @_;
    my $obj = $worker->_api_obj();
    $worker->_git_build(
        {
            id                  => "babl",
            git_checkout_subdir => "babl/git/babl",
            url                 => $worker->_get_gnome_git_url("babl"),
            prefix              => $obj->babl_p,
            use_meson           => 1,
        }
    );
    $worker->_git_build(
        {
            id                  => "gegl",
            git_checkout_subdir => "gegl/git/gegl",
            extra_meson_args    => [ qw# -Dlua=disabled #, ],
            url                 => $worker->_get_gnome_git_url("gegl"),
            prefix              => $obj->gegl_p,
            use_meson           => 1,
        }
    );

    # Override python3_girdir in gexiv2 in order to avoid having
    # to run polkit's pkexec to install files as root/superuser.
    my @gexiv2_girdir_override =
        ( "-Dpython3_girdir=" . ( $obj->gexiv2_p() . "/lib/python3" ), );
    $worker->_git_build(
        {
            id                  => "gexiv2",
            git_checkout_subdir => "gexiv2/git/gexiv2",

            # extra_meson_args    => [ qw# -Dlua=disabled #, ],
            extra_meson_args => [ @gexiv2_girdir_override, ],
            url              => $worker->_get_gnome_git_url("gexiv2"),
            prefix           => $obj->gexiv2_p,
            use_meson        => 1,
        }
    );
    $worker->_git_build(
        {
            id                  => "libmypaint",
            git_checkout_subdir => "libmypaint/git/libmypaint",
            url                 => "https://github.com/mypaint/libmypaint.git",
            prefix              => $obj->mypaint_p,
            use_meson           => 0,
            branch              => "v1.6.1",
            tag                 => "true",
        }
    );
    $worker->_git_build(
        {
            id                  => "mypaint-brushes",
            git_checkout_subdir => "libmypaint/git/mypaint-brushes",
            url       => "https://github.com/Jehan/mypaint-brushes.git",
            prefix    => $obj->mypaint_p,
            use_meson => 0,
            branch    => "v1.3.x",
        }
    );

    my $KEY                    = 'GIMPGITBUILD__BUILD_GIMP_USING_MESON';
    my $BUILD_GIMP_USING_MESON = ( exists( $ENV{$KEY} ) ? $ENV{$KEY} : 1 );

    $worker->_git_build(
        {
            id                   => "gimp",
            extra_configure_args => [ qw# --enable-debug --with-lua=no #, ],
            extra_meson_args     => [ qw# -Dlua=false #, ],
            git_checkout_subdir  => "git/gimp",
            install_before_test  => 1,
            url                  => $worker->_get_gnome_git_url("gimp"),
            prefix               => $obj->gimp_p,
            use_meson            => $BUILD_GIMP_USING_MESON,
            on_failure           => sub {
                my ($args) = @_;
                my $Err = $args->{exception};
                if (   ( $worker->_mode() eq 'clean' )
                    or ( !$BUILD_GIMP_USING_MESON ) )
                {
                    die $Err;
                }
                STDERR->print( $Err, "\n" );
                STDERR->print(<<"EOF");
Meson-using builds of GIMP are known to be error prone. Please try setting
the "$KEY" environment variable to "0", and run gimpgitbuild again, e.g using:

    export $KEY="0"

EOF
                die "Meson build failure";
            },
        }
    );
    return;
}

sub _run_the_mode_on_all_repositories
{
    my ($worker) = @_;

    if ( $worker->_mode() eq 'build' )
    {
        $worker->_override_mode("clean_install");
        $worker->_run_all();
    }
    $worker->_override_mode("");
    $worker->_run_all();

    return;
}

1;

__END__

=head1 NAME

App::gimpgitbuild::API::Worker - common API

=head1 METHODS

=cut
